### ISSUE TITLE
Upgrading from 0.1.2 --> 1.9.4 and Django 1.11 --> 2.2.25 - MSIS9605 related to OpenID?

### ISSUE THREAD

**Issue by 777GE90** [snok/django-auth-adfs#214]
Hello, I am working on upgrading Django to version 2.2.X alongside upgrading our Python to 3.8.

These changes mean I can no longer use version 0.1.2 of django-auth-adfs library.

I have implemented all the required changes to upgrade django-auth-adfs to the latest 1.9.4 version.

However, I have hit a bit of a stumbling block with the new open_id connect code that is introduced in the newer versions of this library.

Essentially, now that I have upgraded, I am getting the following error back from the Microsoft Server 2016:
`MSIS9605%3a+The+client+is+not+allowed+to+access+the+requested+resource.`

I have estbalished that if I comment out the following code in the `build_authorization_endpoint` (config.py) method:
```python3
        # if self._mode == "openid_connect":
        #     query["scope"] = "openid"
        #     if (disable_sso is None and settings.DISABLE_SSO) or disable_sso is True:
        #         query["prompt"] = "login"
        #     if force_mfa:
        #         query["amr_values"] = "ngcmfa"
```

Then the generated authentication URL works fine and I am able to login no problem.

I'm still quite new to ADFS and don't fully understand how openid_connect works, but my organisation has two ADFS instances, one is the legacy instance which uses Active Directory and the other is the newer one using Azure.  We still use the old one and it seems like this openid_connect is related to Azure, so is this plugin wrongly trying to use OpenID when it shouldn't in our instance?

**Comment by JonasKs** [comment]
Oh wow, that's a big jump from `0.1.2`, things are unfortunately bound to break. 😁 

> Ideally, I need a way to force the plugin to stop using openid_connect, I understand it is trying to use openid_connect because it has found a valid config on our server and it uses it by default.

The error says that client is not allowed to access the resource. The lingo can be a bit confusing, so I'd recommend reading [here](https://django-auth-adfs.readthedocs.io/en/latest/oauth2_explained.html#adfs-and-oauth2-lingo-compared). So, the fix isn't really to mess with the code, but to set up ADFS correctly. This [thread ](https://serverfault.com/questions/824506/adfs-error-msis9605-the-client-is-not-allowed-to-access-the-requested-resourc)might help you. 

In other words, I don't want to create a patch to force oauth2. It is legacy, and it would enable _another_ way of configuring this package, which already have way too many paths for me to maintain. This package will most likely shift into only support Azure AD in the future, since I do not have any way of using ADFS (nor want to). I strongly recommend you jump on that train too.

**Comment by JonasKs** [comment]
Hehe, same as me, I don’t have direct access to ADFS. Hence it’s very hard for me to support it anymore. 

I can help you with a branch you can use until you can migrate til Azure AD. That won’t be pushed PyPi and will have to be installed from git though.

**Comment by 777GE90** [comment]
> Hehe, same as me, I don’t have direct access to ADFS. Hence it’s very hard for me to support it anymore.
> 
> I can help you with a branch you can use until you can migrate til Azure AD. That won’t be pushed PyPi and will have to be installed from git though.

I was planning to do the same on our internal Git, but having one here would be neat actually, would possibly help others stuck in similar situations.

**Comment by JonasKs** [comment]
That’s up to you! If you want to fork and link me to it, I can help if needed. I’m not at home, so I won’t be able to help much until later this week.

**Comment by 777GE90** [comment]
> That’s up to you! If you want to fork and link me to it, I can help if needed. I’m not at home, so I won’t be able to help much until later this week.

Sorry got side tracked on another project yesterday, I have already patched this locally now so will leave it.  Have been testing Azure ADFS and seems to work, just having trouble with getting the group membership to come over (they are showing some sort of UUID instead of the group name).  But think I will look into this another time.

Thanks for your help.

**Comment by JonasKs** [comment]
Azure AD -> Application registration -> Token configuration -> sAMAccountName instead of Group ID on `Access` (or all of them).

![bilde](https://user-images.githubusercontent.com/5310116/148936718-ae3e5ce7-fdc4-482f-9f47-66114e6ec1b9.png)

**Comment by JonasKs** [comment]
Closing this issue, but I'll still be notified if you reply. 

Have a good week!

**Comment by 777GE90** [comment]
> Azure AD -> Application registration -> Token configuration -> sAMAccountName instead of Group ID on `Access` (or all of them).
> 
> ![bilde](https://user-images.githubusercontent.com/5310116/148936718-ae3e5ce7-fdc4-482f-9f47-66114e6ec1b9.png)

Awesome, that works :).

You don't happen to know how to get the "winaccountname"? Am I right in understanding that it is not available anymore - seems like I only get the e-mail back now?

**Comment by JonasKs** [comment]
@777GE90  Hmm, no, I'm not sure. We're using `upn`, but not sure if that's 100% the same.

**Comment by JonasKs** [comment]
I have a FastAPI-Azure-Auth package too, and in the tests you can easily compare the tokens from v1 to v2.

In v2 tokens, there are [`prefferred_username`](https://github.com/Intility/fastapi-azure-auth/blob/main/tests/utils.py#L104-L158) which I guess could be used.

**Comment by 777GE90** [comment]
> @777GE90 Hmm, no, I'm not sure. We're using `upn`, but not sure if that's 100% the same.

Ye it differs for us at least, makes the upgrade a bit more work but everything seems very doable now.

> I have a FastAPI-Azure-Auth package too, and in the tests you can easily compare the tokens from v1 to v2.
> 
> In v2 tokens, there are [`prefferred_username`](https://github.com/Intility/fastapi-azure-auth/blob/main/tests/utils.py#L104-L158) which I guess could be used.

Ye I was passing the preferred_username but it doesn't seem to exist in the token, so I am thinking it may not be set in our Azure or it's hidden.

Anyway I'll stop bothering you now lol, thanks for your help :).

**Comment by JonasKs** [comment]
> Ye I was passing the preferred_username but it doesn't seem to exist in the token

This will vary on which token version you configure. FastAPI-Azure-Auth is a much more up2date package, since it's been written _for_ Azure AD. You can see how you change the token version [here](https://intility.github.io/fastapi-azure-auth/single-tenant/azure_setup#step-2---change-token-version-to-v2).
